packages/syft/src/syft/core/node/common/node_manager/bin_obj_manager.py

```diff
@@ -146,12 +154,25 @@ def __setitem__(self, key: UID, value: StorableObject) -> None:
             # name=metadata_dict["name"],
         )
 
+        obj_dataset_relation = self._get_obj_dataset_relation(key)
+        if obj_dataset_relation:
+            # Create a object dataset relationship for the new object
+            obj_dataset_relation = BinObjDataset(
+                id=obj_dataset_relation.id,
+                name=obj_dataset_relation.name,
+                obj=bin_obj.id,
+                dataset=obj_dataset_relation.dataset,
+                dtype=obj_dataset_relation.dtype,
+                shape=obj_dataset_relation.shape,
+            )
+
         if self.__contains__(key):
             self.delete(key)
```

Comment:
Since the BinObject is deleted, the corresponding BinObjDataset is also deleted due to cascading effect. Thus we need to recreate the BinObjDataset for the new BinObject being created.